### PR TITLE
Update to Cython 3.0.0

### PR DESCRIPTION
**Pull request by vyasr**
This PR bumps the pinning and fixes the new warnings that appeared

### CODE REVIEW

File: python/kvikio/_lib/libnvcomp_ll.pyx [hunk 1]
```diff
@@ -76,7 +76,7 @@ class nvCompBatchAlgorithm(ABC):
         self,
         size_t batch_size,
         size_t max_uncompressed_chunk_bytes,
-    ) -> (nvcompStatus_t, size_t):
+    ) -> tuple[nvcompStatus_t, size_t]:
```
Comment (Alexey-Kamenev) [inline python/kvikio/_lib/libnvcomp_ll.pyx:79]: So I guess this is where the dependency on Python 3.9+ is introduced (PEP-585), right? Up until now, `kvikio` did not use any Python feature which would specifically require 3.9+.

Comment (vyasr) [reply python/kvikio/_lib/libnvcomp_ll.pyx:79]: That's an interesting point I didn't consider. I believe in this particular context this change is actually backwards compatible because using tuple directly has been supporting since Python 3.7, it [just wasn't fully featured](https://peps.python.org/pep-0585/#implementation):

> Usefulness of this syntax before [PEP 585](https://peps.python.org/pep-0585) is limited as external tooling like Mypy does not recognize standard collections as generic. Moreover, certain features of typing like type aliases or casting require putting types outside of annotations, in runtime context.

I would just need to add a `from __future__ import annotations` to this file. I don't know if Cython 3.0.0 relies on Python 3.9+ functionality to recognize this type, but it wouldn't matter much anyway here since this function isn't a `cdef` function. If you prefer I could switch to using `typing.tuple` for now instead though. That may eventually be deprecated, but not for a couple of years at least, and it would be fully backwards compatible.

**Review by raydouglass — APPROVED** [review]

File: python/kvikio/_lib/nvcomp_cxx_api.pxd [hunk 2]
```diff
@@ -73,7 +73,6 @@ cdef extern from "nvcomp/nvcompManager.hpp" namespace 'nvcomp':
             size_t uncompressed_buffer_size) except +
         nvcompStatus_t* get_status() const
         CompressionConfig(CompressionConfig& other)
-        CompressionConfig& operator=(CompressionConfig&& other) except +
```
Comment (vuule) [inline python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: why are these removed?

Comment (vyasr) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: Cython doesn't actually support rvalue reference parameters. It never did, but previously it wouldn't warn about it. Now it does.

Note that despite this, you can still get the desired behavior by writing Cython code with an explicit move `x ==move(CompressionConfig())`. Cython will generate the C++ code assuming it'll use one of the other operators for an lvalue, but then the C++ compiler will use the rvalue version instead.

Comment (vuule) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: The only issue might be that there's zero visibility of the rvalue versions on the Cython side. Should we leave a comment that a move is applicable when calling operator=?

Comment (vyasr) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: We could do that. We may need a more comprehensive strategy for this though since it's a problem that affects all of RAPIDS. There are multiple cases where we have functions that accept rvalues that we may want to leverage in Cython even though Cython doesn't think it knows how. @shwina do you have thoughts on how best to handle this?

Comment (shwina) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: I don't have any fantastic suggestions. Cython doesn't really know about rvalues generally and getting it to generate the right code involves using explicit `move()` calls.

Agree with @vuule - perhaps we just leave a comment?

Comment (jakirkham) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: Proposing such a change below. Maybe we can iterate on that change?

File: python/kvikio/_lib/nvcomp_cxx_api.pxd [hunk 3]
```diff
@@ -73,7 +73,6 @@ cdef extern from "nvcomp/nvcompManager.hpp" namespace 'nvcomp':
...
         CompressionConfig& operator=(const CompressionConfig& other) except +
```
Comment (jakirkham) [inline python/kvikio/_lib/nvcomp_cxx_api.pxd:76]:
```suggestion
        CompressionConfig& operator=(const CompressionConfig& other) except +
        # Commented as Cython doesn't support rvalues, but
        # a user can call `move` with the existing operator and generate correct C++ code
        # xref: https://github.com/cython/cython/issues/1445
        # CompressionConfig& operator=(CompressionConfig&& other) except +
```

File: python/kvikio/_lib/nvcomp_cxx_api.pxd [hunk 4]
```diff
@@ -82,7 +81,6 @@ cdef extern from "nvcomp/nvcompManager.hpp" namespace 'nvcomp':
         DecompressionConfig(PinnedPtrPool[nvcompStatus_t]& pool) except +
         nvcompStatus_t* get_status() const
         DecompressionConfig(DecompressionConfig& other)
-        DecompressionConfig& operator=(DecompressionConfig&& other) except +
         DecompressionConfig& operator=(const DecompressionConfig& other) except +
```
Comment (jakirkham) [inline python/kvikio/_lib/nvcomp_cxx_api.pxd:84]:
```suggestion
        CompressionConfig& operator=(const CompressionConfig& other) except +
        # Commented as Cython doesn't support rvalues, but
        # a user can call `move` with the existing operator and generate correct C++ code
        # xref: https://github.com/cython/cython/issues/1445
        # DecompressionConfig& operator=(DecompressionConfig&& other) except +
```

**Review by jakirkham — COMMENTED** [review]
Thanks Vyas! 🙏

Proposed a comment that may capture what we are after with rvalues

File: python/kvikio/_lib/libnvcomp_ll.pyx (hunk 1 above)
Comment (Alexey-Kamenev) [reply python/kvikio/_lib/libnvcomp_ll.pyx:87]: Apologies for not responding sooner.  I think adding `from __future__ import annotations` is a good idea. The package already requires 3.9+ anyway, so anyone willing to use it with 3.8 will be aware of potential compatibility issues.

File: python/kvikio/_lib/nvcomp_cxx_api.pxd (hunk 2 above)
Comment (vyasr) [reply python/kvikio/_lib/nvcomp_cxx_api.pxd:76]: Done.

**Comment by jakirkham** [comment]
@Alexey-Kamenev @vuule , could you please review the latest changes?

**Review by Alexey-Kamenev — APPROVED** [review]
Looks good to me!

**Review by vuule — APPROVED** [review]

**Review by jakirkham — APPROVED** [review]
